components/file_system/src/rate_limiter.rs

```diff
@@ -1,148 +1,527 @@
-        match io_op {
-            IOOp::Read => self.read[io_type as usize].load(Ordering::Relaxed),
-            IOOp::Write => self.write[io_type as usize].load(Ordering::Relaxed),
+    pub fn reset(&self) {
+        for i in 0..IOType::COUNT {
+            self.read_bytes[i].store(0, Ordering::Relaxed);
+            self.write_bytes[i].store(0, Ordering::Relaxed);
+        }
+    }
+}
+
+macro_rules! do_sleep {
+    ($duration:expr, sync) => {
+        std::thread::sleep($duration)
+    };
+    ($duration:expr, async) => {
+        tokio::time::delay_for($duration).await
+    };
+}
+
+const DEFAULT_REFILL_PERIOD: Duration = Duration::from_millis(50);
+
+/// Limit total IO flow below provided threshold by throttling lower-priority IOs.
+/// Rate limit is disabled when total IO threshold is set to zero.
+#[derive(Debug)]
+struct PriorityBasedIORateLimiter {
+    // IO amount passed through within current epoch
+    bytes_through: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    // Maximum IOs permitted within current epoch
+    bytes_per_epoch: [CachePadded<AtomicUsize>; IOPriority::COUNT],
+    protected: Mutex<PriorityBasedIORateLimiterProtected>,
+}
+
+#[derive(Debug)]
+struct PriorityBasedIORateLimiterProtected {
+    next_refill_time: Instant,
+    // IOs that are can't be fulfilled in current epoch
+    pending_bytes: [usize; IOPriority::COUNT],
+    // Used to smoothly update IO budgets
+    history_epoch_count: usize,
+    history_bytes: [usize; IOPriority::COUNT],
+}
+
+impl PriorityBasedIORateLimiterProtected {
+    fn new() -> Self {
+        PriorityBasedIORateLimiterProtected {
+            next_refill_time: Instant::now_coarse() + DEFAULT_REFILL_PERIOD,
+            pending_bytes: [0; IOPriority::COUNT],
+            history_epoch_count: 0,
+            history_bytes: [0; IOPriority::COUNT],
         }
     }
+}
 
+/// Actual implementation for requesting IOs from PriorityBasedIORateLimiter.
+/// An attempt will be recorded first. If the attempted amount exceeds the available quotas of
+/// current epoch, the requester will register itself and sleep until next epoch.
+macro_rules! request_imp {
+    ($limiter:ident, $priority:ident, $amount:ident, $mode:tt) => {{
+        let priority_idx = $priority as usize;
+        let cached_bytes_per_refill =
+            $limiter.bytes_per_epoch[priority_idx].load(Ordering::Relaxed);
+        if cached_bytes_per_refill == 0 {
+            return $amount;
+        }
+        let amount = std::cmp::min($amount, cached_bytes_per_refill);
+        let bytes_through =
+            $limiter.bytes_through[priority_idx].fetch_add(amount, Ordering::AcqRel) + amount;
+        if bytes_through <= cached_bytes_per_refill {
+            return amount;
+        }
+        let now = Instant::now_coarse();
+        let mut wait = Duration::from_millis(0);
+        // hold a snapshot ticket of pending bytes
+        let pending_snapshot = {
```

Comment:
what do you mean by `snapshot`?

Reply:
Added a comment. Basically we use the `pending_bytes` value at this point to deduce when this slice of newly added pending request can actually be served.